File: ios/Airgap/ProviderHarness.swift
import Foundation

enum ProviderHarnessError: Error {
  case manifestMissing
  case scenarioArgumentMissing
  case unknownScenario
  case unsupportedPlatform
}

struct ProviderHarnessManifest: Decodable {
  let schemaVersion: Int
  let scenarios: [ProviderHarnessScenario]
}

struct ProviderHarnessScenario: Decodable {
  let id: String
  let platforms: [String]
  let capabilities: [String: ProviderHarnessCapability]
  let generation: ProviderHarnessGeneration
  let downloadProgress: [Double]?
}

struct ProviderHarnessCapability: Decodable {
  let state: String
  let contextSize: Int?
  let modelIdentity: String?
  let reason: String?
}

struct ProviderHarnessGeneration: Decodable {
  let tokens: [String]?
  let text: String?
  let error: String?
}

struct ProviderHarnessResolvedScenario {
  let id: String
  let capability: ProviderHarnessCapability
  let generation: ProviderHarnessResolvedGeneration
  let downloadProgress: [Double]
}

struct ProviderHarnessResolvedGeneration {
  let tokens: [String]
  let text: String?
  let error: String?
}

enum ProviderHarness {
  static let argumentName = "-AirgapProviderScenario"

  static func requestedScenarioName(arguments: [String] = CommandLine.arguments) -> String? {
    guard
      let argumentIndex = arguments.firstIndex(of: argumentName),
      arguments.indices.contains(argumentIndex + 1)
    else {
      return nil
    }
    let name = arguments[argumentIndex + 1].trimmingCharacters(in: .whitespacesAndNewlines)
    return name.isEmpty ? nil : name
  }

  static func load(
    bundle: Bundle = .main,
    arguments: [String] = CommandLine.arguments
  ) throws -> ProviderHarnessResolvedScenario {
    guard let scenarioName = requestedScenarioName(arguments: arguments) else {
      throw ProviderHarnessError.scenarioArgumentMissing
    }
    guard let url = bundle.url(forResource: "provider-scenarios", withExtension: "json") else {
      throw ProviderHarnessError.manifestMissing
    }

    return try load(
      data: Data(contentsOf: url),
      scenarioName: scenarioName,
      platform: "ios"
    )
  }

  static func load(
    data: Data,
    scenarioName: String,
    platform: String
  ) throws -> ProviderHarnessResolvedScenario {
    let manifest = try JSONDecoder().decode(ProviderHarnessManifest.self, from: data)
    let normalizedName = scenarioName.trimmingCharacters(in: .whitespacesAndNewlines)
    let normalizedPlatform = platform.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()

    guard let scenario = manifest.scenarios.first(where: { $0.id == normalizedName }) else {
      throw ProviderHarnessError.unknownScenario
    }
    guard
      scenario.platforms.contains(normalizedPlatform),
      let capability = scenario.capabilities[normalizedPlatform]
    else {
      throw ProviderHarnessError.unsupportedPlatform
    }

    return ProviderHarnessResolvedScenario(
      id: scenario.id,
      capability: capability,
      generation: ProviderHarnessResolvedGeneration(
        tokens: scenario.generation.tokens ?? [],
        text: scenario.generation.text,
        error: scenario.generation.error
      ),
      downloadProgress: scenario.downloadProgress ?? []
    )
  }
}

File: android/app/src/main/java/com/airgap/inference/ProviderHarness.kt
package com.airgap.inference

import org.json.JSONObject

class ProviderHarnessException(message: String) : IllegalArgumentException(message)

data class ProviderHarnessCapability(
    val state: String,
    val contextSize: Int?,
    val modelIdentity: String?,
    val reason: String?,
)

data class ProviderHarnessGeneration(
    val tokens: List<String>,
    val text: String?,
    val error: String?,
)

data class ProviderHarnessScenario(
    val id: String,
    val platforms: List<String>,
    val capabilities: Map<String, ProviderHarnessCapability>,
    val generation: ProviderHarnessGeneration,
    val downloadProgress: List<Double>,
)

data class ProviderHarnessResolvedScenario(
    val id: String,
    val capability: ProviderHarnessCapability,
    val generation: ProviderHarnessGeneration,
    val downloadProgress: List<Double>,
)

object ProviderHarness {
    fun load(
        json: String,
        scenarioName: String,
        platform: String = "android",
    ): ProviderHarnessResolvedScenario {
        val normalizedName = scenarioName.trim()
        val normalizedPlatform = platform.trim().lowercase()
        val scenario = loadAll(json).firstOrNull { it.id == normalizedName }
            ?: throw ProviderHarnessException("provider_harness_scenario_unknown")
        val capability = scenario.capabilities[normalizedPlatform]
            ?: throw ProviderHarnessException("provider_harness_platform_unsupported")
        if (!scenario.platforms.contains(normalizedPlatform)) {
            throw ProviderHarnessException("provider_harness_platform_unsupported")
        }

        return ProviderHarnessResolvedScenario(
            id = scenario.id,
            capability = capability,
            generation = scenario.generation,
            downloadProgress = scenario.downloadProgress,
        )
    }

    fun loadAll(json: String): List<ProviderHarnessScenario> {
        val root = JSONObject(json)
        val scenarios = root.getJSONArray("scenarios")
        return (0 until scenarios.length()).map { index ->
            parseScenario(scenarios.getJSONObject(index))
        }
    }

    private fun parseScenario(json: JSONObject): ProviderHarnessScenario {
        val platforms = json.getJSONArray("platforms").let { values ->
            (0 until values.length()).map(values::getString)
        }
        val capabilitiesJson = json.getJSONObject("capabilities")
        val capabilities = capabilitiesJson.keys().asSequence().associateWith { platform ->
            parseCapability(capabilitiesJson.getJSONObject(platform))
        }
        val generationJson = json.getJSONObject("generation")
        val tokens = if (generationJson.has("tokens")) {
            generationJson.getJSONArray("tokens").let { values ->
                (0 until values.length()).map(values::getString)
            }
        } else {
            emptyList()
        }
        val progress = if (json.has("downloadProgress")) {
            json.getJSONArray("downloadProgress").let { values ->
                (0 until values.length()).map(values::getDouble)
            }
        } else {
            emptyList()
        }

        return ProviderHarnessScenario(
            id = json.getString("id"),
            platforms = platforms,
            capabilities = capabilities,
            generation = ProviderHarnessGeneration(
                tokens = tokens,
                text = generationJson.optionalString("text"),
                error = generationJson.optionalString("error"),
            ),
            downloadProgress = progress,
        )
    }

    private fun parseCapability(json: JSONObject) = ProviderHarnessCapability(
        state = json.getString("state"),
        contextSize = if (json.has("contextSize")) json.getInt("contextSize") else null,
        modelIdentity = json.optionalString("modelIdentity"),
        reason = json.optionalString("reason"),
    )

    private fun JSONObject.optionalString(name: String): String? =
        if (has(name) && !isNull(name)) getString(name) else null
}
